### DISCUSSION TITLE
Custom interface API requests

### DISCUSSION THREAD

**Discussion by dallashuggins** [directus/directus#8534] [Archive]
Is there any built-in functionality for making Directus API requests in a custom interface?

From reviewing some of the discussions, it seems as though Directus provides an `api` that you can provide as the `system`, which I cannot get working. 

Import: `import { api } from 'directus';`
In `defineComponent`: `inject: ['system'],`
In setup function: `provide('system', { api })`

Whenever I import the api I get the following error when I try to access the CMS, where it'll say it can't load custom interfaces and throw a TypeError:
```
TypeError: Error resolving module specifier “crypto”. Relative module specifiers must start with “./”, “../” or “/”.
```

When building the app, I get many variations of this error: `⠧ Building Directus extension...preferring built-in module 'util' over local alternative at '/Users/dallas/node_modules/util/util.js', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning`. When starting the app, it says many of the packages aren't resolved (i.e., `'crypto' is imported by extensions/interfaces/price-conversion/index.js, but could not be resolved – treating it as an external dependency`). 

Any ideas what I may be doing wrong?

Also, what is the `system` prop for interfaces (`defineInterface`)? I was wondering if the prop was supposed to inject the system automatically so you can make API requests, similar to the discussion [here](https://github.com/directus/directus/discussions/4659). It always comes back undefined when I use that prop. It is shown in the [Typescript definition](https://github.com/directus/directus/blob/20355fee5eba514dd75565f60269311187010c66/app/src/interfaces/types.ts#L5-L18) linked in the [docs](https://docs.directus.io/guides/interfaces/), but when I switch the branch to `main` it 404s on that page, and I cannot find any information otherwise on what it does.

**Comment by nickrum** [comment]
You can't do `import { api } from 'directus';` inside an App extension because the `directus` package is actually the Directus API server. Your browser is throwing an error because it is trying to import a Node.js-only library (`crypto` in this case).

We replaced the `system` inject with multiple more specific injects and exposed composables to use them, so you should be able to do something like this:

```js
import { useApi } from '@directus/extensions-sdk';

...

setup() {
  const api = useApi();

  ...
}
```

> Also, what is the `system` prop for interfaces (`defineInterface`)?

In this case, `system` is not a prop but a boolean flag used for built in system interfaces.

**Reply by dallashuggins** [reply]
Awesome, thank you so much! That makes sense, and I confirmed `useApi` worked perfectly. 🙌

**Reply by rijkvanzanten** [reply]
cc @alesvaupotic
